base/osutils.jl

```diff
@@ -2,59 +2,95 @@
 
 const OS_NAME = ccall(:jl_get_OS_NAME, Any, ())
 
-function is_unix(os::Symbol)
-    if (os==:Windows) return false;
-    elseif (os==:Linux) return true;
-    elseif (os==:FreeBSD) return true;
-    elseif (os==:Darwin) return true;
-    else throw(ArgumentError("unknown operating system, $(repr(os))"))
-    end
-end
+"""
+    is_unix([os])
 
-function _os_test(qm,ex,test)
-    @assert qm == :?
-    @assert isa(ex,Expr)
-    @assert ex.head == :(:)
-    @assert length(ex.args) == 2
-    if test
-        return esc(ex.args[1])
+Predicate for testing if the OS provides a Unix-like interface.
+See documentation :ref:`Handling Operating System Variation <man-handling-operating-system-variation>`\ .
+"""
+function is_unix(os::Symbol = OS_NAME)
+    if is_windows(os)
+        return false
+    elseif is_linux(os) || is_bsd(os)
+        return true
     else
-        return esc(ex.args[2])
+        throw(ArgumentError("unknown operating system \"$os\""))
```

Comment:
I don't see any reason `pure` functions can't throw exceptions (preferably as long as they always throw the exception). fwiw, I don't actually enforce that the call in `@static` must be pure, but left that as a future exercise.

In a few places, I used `@static` mostly as a way of saving a few bytes in the sysimg. There's only a few contexts in which it makes a significant difference right now (`ccall` needs a constant name that is defined on the current system)